src/GafferScene/EvaluateLightLinks.cpp

```diff
@@ -0,0 +1,127 @@
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include <cstdio>
+
+#include "IECore/CompoundData.h"
+
+#include "GafferScene/EvaluateLightLinks.h"
+#include "GafferScene/SetAlgo.h"
+
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+IE_CORE_DEFINERUNTIMETYPED( EvaluateLightLinks );
+
+EvaluateLightLinks::EvaluateLightLinks( const std::string &name )
+	:	SceneProcessor( name )
+{
+	m_lightLinkAttrName = "gaffer:linkedLights";
+
+	// Fast pass-throughs for the things we don't alter.
+	outPlug()->boundPlug()->setInput( inPlug()->boundPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+	outPlug()->setNamesPlug()->setInput( inPlug()->setNamesPlug() );
+	outPlug()->setPlug()->setInput( inPlug()->setPlug() );
+	outPlug()->objectPlug()->setInput( inPlug()->objectPlug() );
+	outPlug()->childNamesPlug()->setInput( inPlug()->childNamesPlug() );
+	outPlug()->globalsPlug()->setInput( inPlug()->globalsPlug() );
+}
+
+EvaluateLightLinks::~EvaluateLightLinks()
+{
+}
+
+void EvaluateLightLinks::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneProcessor::affects( input, outputs );
+
+	if( input == inPlug()->attributesPlug() || input == inPlug()->setPlug() || input == inPlug()->globalsPlug() )
```

Comment:
I like the `SetAlgo::affectsSetExpression` idea and have added it in one of the commits I'll be pushing in a bit. Could you have a look? One thing I'm not sure of is whether it's ok to just check the `Direction` and by doing so allowing all inputs to affect the outputs as opposed to what the node did before by looking at `inPlug()` only. For `EvaluateLightLinks` it doesn't matter, but it might for a node that we come up with in the future? Without checking the `Direction` I got errors because the `attributesPlug` of the `outPlug` would then depend on itself. 
We could just pass in the parent explicitly, but I kinda liked the clean function signature. Do you see a problem with potentially depending on multiple inputs? @danieldresser pointed out that in the cases in which we have multiple inputs we might actually want to depend on all of them anyway, because a merge node for example would use all the data that is flowing in.